client/tx_watcher.go

```
@@ -0,0 +1,113 @@
+func NewTxWatcher(client client, clientTimeout time.Duration, maxRetries int, waitBetweenRetries time.Duration, percentageIncrease float64) *TxWatcher {
+	return &TxWatcher{
+		client:             client,
+		clientTimeout:      clientTimeout,
+		maxRetries:         maxRetries,
+		waitBetweenRetries: waitBetweenRetries,
+		percentageIncrease: percentageIncrease,
+	}
+}
+
+// WatchableTransaction represents a transaction that the txwatcher can keep track of
+type WatchableTransaction func(gasPrice *big.Int) (*types.Transaction, error)
+
+// EnsureTransactionSuccess keeps track of the transaction on blockchain.
+// If it sees that the transaction has not appeared on BC, it will try again and increase the gas price.
+func (tw *TxWatcher) EnsureTransactionSuccess(wt WatchableTransaction, initialGasPrice *big.Int) (chan *types.Transaction, func(), chan error) {
+	stop := make(chan struct{})
+	once := sync.Once{}
+	cancel := func() {
+		once.Do(func() {
+			close(stop)
+		})
+	}
+	errChan := make(chan error)
+	txChan := make(chan *types.Transaction)
+
+	go func() {
+		defer close(txChan)
+		defer close(errChan)
+		tx, err := tw.watchTx(wt, initialGasPrice, stop)
+		txChan <- tx
+		errChan <- err
+	}()
+
+	return txChan, cancel, errChan
+}
+
+func (tw *TxWatcher) watchTx(wt WatchableTransaction, initialGasPrice *big.Int, stop <-chan struct{}) (*types.Transaction, error) {
+	for i := 1; i <= tw.maxRetries; i++ {
+		// if sending fails just notify the caller
+		gasPrice := tw.calculateGasPrice(initialGasPrice, i)
+		tx, err := wt(gasPrice)
+		if err != nil {
+			return nil, err
+		}
+
+		ctx, c := context.WithTimeout(context.Background(), tw.clientTimeout)
+		defer c()
+		pulledTx, _, err := tw.client.TransactionByHash(ctx, tx.Hash())
+		if err != nil {
+			log.Warn().Err(err).Msgf("transaction %q not visible on BC, will retry", tx.Hash())
+			continue
+		}
+
+		if pulledTx != nil {
+			return pulledTx, nil
+		}
+
+		select {
+		case <-stop:
+			return nil, nil
```

Should we return an error here that it was interrupted from outside?